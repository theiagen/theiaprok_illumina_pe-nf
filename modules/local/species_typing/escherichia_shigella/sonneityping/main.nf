process SONNEITYPER {
    tag "$meta.id"
    label "process_medium"

    container "us-docker.pkg.dev/general-theiagen/staphb/mykrobe:0.12.1-sonneityping"

    input:
    tuple val(meta), path(reads)
    val ont_data

    output:
    tuple val(meta), path("*.mykrobe.csv"), optional: true, emit: mykrobe_report_csv
    tuple val(meta), path("*.mykrobe.json"), optional: true, emit: mykrobe_report_json
    tuple val(meta), path("*.sonneityping.tsv"), optional: true, emit: sonneityping_final_report
    tuple val(meta), path("SPECIES.txt"), emit: sonneityping_species
    tuple val(meta), path("FINAL_GENOTYPE.txt"), emit: sonnietyping_final_genotype
    tuple val(meta), path("GENOTYPE_NAME.txt"), emit: sonneityping_genotype_name
    tuple val(meta), path("CONFIDENCE.txt"), emit: genotype_confidence
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def ont_flag = ont_data ? "--ont" : ""
    def reads_input = reads.collect{ it.toString() }.join(' ')
    
    """
    # Print and save versions
    mykrobe --version | sed 's|mykrobe v||g' | tee MYKROBE_VERSION.txt
    
    # Run Mykrobe on the input read data
    mykrobe predict \\
        -t ${task.cpus} \\
        --sample ${prefix} \\
        --species sonnei \\
        --format json_and_csv \\
        --out ${prefix}.mykrobe \\
        ${ont_flag} \\
        --seq ${reads_input} \\
        ${args}
    
    # use sonneityping script to produce final TSV; alleles.txt is required input for human-readable genotype names
    python /sonneityping/parse_mykrobe_predict.py \\
        --jsons ${prefix}.mykrobe.json \\
        --alleles /sonneityping/alleles.txt \\
        --prefix ${prefix}.sonneityping
    
    if [ -f ${prefix}.sonneityping_predictResults.tsv ]; then
        echo "DEBUG: sonneityping produced expected output file"
        # rename output TSV to something prettier
        mv -v ${prefix}.sonneityping_predictResults.tsv ${prefix}.sonneityping.tsv

        # Parse the output files to create final reports
        sonneityper_parser.py --sonnei_tsv ${prefix}.sonneityping.tsv
    else 
        echo "Error: sonneityping did not produce expected output file. Check mykrobe logs."
        touch SPECIES.txt
        touch FINAL_GENOTYPE.txt
        touch GENOTYPE_NAME.txt
        touch CONFIDENCE.txt
    fi
    

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mykrobe: \$(mykrobe --version | sed 's|mykrobe v||g')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.mykrobe.csv
    touch ${prefix}.mykrobe.json
    touch ${prefix}.sonneityping.tsv
    touch SPECIES.txt
    touch FINAL_GENOTYPE.txt
    touch GENOTYPE_NAME.txt
    touch CONFIDENCE.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
       mykrobe: \$(mykrobe --version | sed 's|mykrobe v||g')
    END_VERSIONS
    """
}