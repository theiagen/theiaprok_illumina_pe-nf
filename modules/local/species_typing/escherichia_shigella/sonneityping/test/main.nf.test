nextflow_process {
    name "Test Process SONNEITYPER"
    script "../main.nf"
    process "SONNEITYPER"
    
    test("sonneityper - test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = false
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.mykrobe_report_csv,
                    process.out.mykrobe_report_json,
                    process.out.sonneityping_final_report,
                    process.out.sonneityping_species,
                    process.out.sonneityping_final_genotype,
                    process.out.sonneityping_genotype_name,
                    process.out.genotype_confidence,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("sonneityper - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = false
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}