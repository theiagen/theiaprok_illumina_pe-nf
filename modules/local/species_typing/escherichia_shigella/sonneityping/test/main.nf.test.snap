{
    "sonneityper - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.mykrobe.csv:md5,242c9b02404c2aba9a380ed95592b0be"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.mykrobe.json:md5,bcaf61c7c2b7262deaefee4a251c9127"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.sonneityping.tsv:md5,897dd1ce907df7e1acdad381a89329a4"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "SPECIES.txt:md5,bde357f0a8af0bea5b4dbf47f14ff988"
                ]
            ],
            null,
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "GENOTYPE_NAME.txt:md5,d4cd0dabcf4caa22ad92fab40844c786"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "CONFIDENCE.txt:md5,d4cd0dabcf4caa22ad92fab40844c786"
                ]
            ],
            [
                "versions.yml:md5,076b2c1451625aecb218634e9a61337a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T20:53:16.067579207"
    },
    "sonneityper - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.mykrobe.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.mykrobe.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.sonneityping.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SPECIES.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "FINAL_GENOTYPE.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "GENOTYPE_NAME.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "CONFIDENCE.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    "versions.yml:md5,296cf002785168be7059459ab971b893"
                ],
                "genotype_confidence": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "CONFIDENCE.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "mykrobe_report_csv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.mykrobe.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "mykrobe_report_json": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.mykrobe.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sonneityping_final_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.sonneityping.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sonneityping_genotype_name": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "GENOTYPE_NAME.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sonneityping_species": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SPECIES.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sonnietyping_final_genotype": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "FINAL_GENOTYPE.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,296cf002785168be7059459ab971b893"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T20:53:29.3410081"
    }
}